Kuadrant/wasm-shim
# prefix based path matching rule

### what

Support prefix based path matching

The following config will generate descriptors for the request `GET /v1/service/my-service/something`

```yaml
failure_mode_deny: true
rate_limit_policies:
  - name: toystore
    rate_limit_domain: toystore-app
    upstream_cluster: rate-limit-cluster
    hostnames: ["*.toystore.com"]
    gateway_actions:
      - rules:
          - paths: ["/v1/service*"]   <- prefix based path matching rule
            methods: ["GET"]
        configurations:
          - actions:
            - generic_key:
                descriptor_key: admin
                descriptor_value: "1"
```
Required by https://github.com/Kuadrant/kuadrant-controller/issues/202

Additionally:
* use `latest` limitador version in the dev environment (based on docker compose)
* Adapt dev environment limits file to the latest syntax
* FIX metadata action type when the key is not found

### verification steps

Run dev env

```
make development
```
The wasm config in the envoy for the `*.b.com` workload look like this
```json
{
  "name": "rlp-b",
  "rate_limit_domain": "rls-domain-b",
  "upstream_cluster": "limitador",
  "hostnames": ["*.b.com"],
  "gateway_actions": [
    {
      "rules": [
        {
          "paths": ["/not/exising/path", "/any*"]           <- prefix based path matching
        }
      ],
      "configurations": [
        {
          "actions": [
            {
              "generic_key": {
                "descriptor_key": "some_generic_key",
                "descriptor_value": "some_generic_value"
              }
            }
          ]
        }
      ]
    }
  ]
}
```

Check that the request 
```
curl -H "Host: test.b.com" http://127.0.0.1:18000/anything
```
is generating descriptor entries. It can easily be seen in the lmitador's logs or checking that after 5 request, the requests are rate limited with 429 responses. If it is rate limited, descriptors must have been generated. Currently the limit file looks like this

```yaml
 - namespace: rls-domain-b                         
   max_value: 5                                    
   seconds: 10                                     
   conditions:                                     
   ⦙ - "some_generic_key == 'some_generic_value'"  
```
